src/mscorlib/shared/System/Buffers/TlsOverPerCoreLockedStacksArrayPool.cs

```diff
@@ -177,20 +189,29 @@ public override void Return(T[] array, bool clearArray = false)
                 // if there was a previous one there, push that to the global stack.  This
                 // helps to keep LIFO access such that the most recently pushed stack will
                 // be in TLS and the first to be popped next.
-                T[][] tlsBuckets = t_tlsBuckets;
+                Bucket[] tlsBuckets = t_tlsBuckets;
                 if (tlsBuckets == null)
                 {
-                    t_tlsBuckets = tlsBuckets = new T[NumBuckets][];
-                    tlsBuckets[bucketIndex] = array;
+                    t_tlsBuckets = tlsBuckets = new Bucket[NumBuckets];
+                    tlsBuckets[bucketIndex].Buffer = array;
+                    if (TrimBuffers)
+                    {
+                        s_allTlsBuckets.Add(tlsBuckets, null);
+                        Gen2GcCallback.Register(Gen2GcCallbackFunc, this);
+                    }
                 }
                 else
                 {
-                    T[] prev = tlsBuckets[bucketIndex];
-                    tlsBuckets[bucketIndex] = array;
+                    T[] prev = tlsBuckets[bucketIndex].Buffer;
+                    tlsBuckets[bucketIndex].Buffer = array;
+                    if (TrimBuffers)
+                    {
+                        tlsBuckets[bucketIndex].Ticks = (uint)Environment.TickCount;
```

Comment:
I'm surprised we're not seeing an impact on return speed, as this is effectively adding a TickCount call to every return.  Is TickCount so fast as to be imperceptible?  Is that true on both Windows and Unix?

Reply:
On Windows it didn't impact the numbers in a perceptible way. Unix does do some calculating here, not sure what options we would have outside of not tracking ticks on TLS. Open to ideas.

Reply:
TickCount is pretty fast at least on Windows (MeasureIt says it is < the cost of 3 call-returns).     I don't know about Linux.

 I too am worried about this code path, as we have made this method substantially more expensive than it was for a common case of allocating and releasing over and over.   

The alternative of course, is to have the Trim() callback 'age' the entries, and the only thing you have to do on the hot path is to reset the age to 0.    GIven that we are enumerating every entry during these Trim() operations anyway, this seems to be the way to go (thus Ticks become 'age' and every time Trim() happens the age goes up.  

Thinking about it, having 'Trim' based off the Gen2 GCs is actually not that great because it means you free things at EXACTLY the wrong time (you have to wait for another Gen2 GC to actually get the memory back).   Having a timer that goes off say every 10 seconds to update the 'age' fields and free things if possible seems like something we should consider.   Interestingly the way we can be more aggressive is to simply change the Trim() callback interval.

(Note that it would be good if we actually used GC.WaitForFullGCApproach method (which was really intended for cache management like this), but the way its designed, you need to burn a thread to use it) so I am not suggesting that here...

Reply:
Having said all this, is I am not saying that the current approach is bad.   Just that we should consider both.   It *seems* like the 'Trim() ages entries' approach is worth careful consideration...

Reply:
Since we're adding GC API anyway have we considered offering a (private) callback version of GC.WaitForFullGCApproach that doesn't tie up a thread?

Reply:
> The alternative of course, is to have the Trim() callback 'age' the entries, and the only thing you have to do on the hot path is to reset the age to 0. 

Given that the Gen2 isn't regular I'm not sure how you would accurately track that. Either we'd end up killing too quickly or we'd end up waiting an extra gen 2 GC above what we're already doing to get the memory back.

> you have to wait for another Gen2 GC to actually get the memory back

Yes, and this is why we try to get ahead of the GC actually hitting the high water mark. (e.g. the .90/.70 marks)

> Having a timer that goes off say every 10 seconds 

I worry about having too many callbacks happening. Right now we're getting reports of collapse after running for a day. I sort of like being driven by Gen2s given that bar.

> the way its designed, you need to burn a thread 

Yeah, that was a primary design consideration for me. Didn't want to introduce another thread or try backpacking on some thread we've not backpacked on before.

Reply:
> I worry about having too many callbacks happening. Right now we're getting reports of collapse after running for a day. I sort of like being driven by Gen2s given that bar.

Note that architecture of 'Trim does the aging' is decoupled from what actually kicks off Trim().  It is 'the 'Tick()', and we can define it however we like.  It can be simply GC2 count, but it can also be pure time (e.g. 10 seconds), or hybrids (GC2 or 10 seconds whichever is shorter).   

To make things concrete here is a SIMPLE implementation.   ticks -> age and age starts at 0 and gets reset to zero every time the entry is returned.    Trim first adds the time since the last Trim to the age, and then does what it does today (based on memory pressure remove entries).   Notice we have also remove the wrap-around problem with Ticks.   It also makes Return faster since there is no called to Environment.GetTicks.   You can even save memory by making ticks be in a byte (in parallel arrays)  (A byte is plenty because you only a modest number of distinct ages (because you only use it to make coarse selection)).     

In the SIMPLEST implementation, you make Tick go off every say 10 seconds.   Note that it can be GC aware easily (without Gen2GCCallback) by simply calling GCCount to get he number of Gen2 GCs. 

If you feel strongly that things could be completely out of control even within 10 seconds, then you could add back in the callback on Gen2 GCs, but I would not do that (it is more complex).   You can also decide to dispense with the Timer and just use Gen2s (you just have to note the time since the last Gen2.   The only down side here is that unless you jetteson EVERYTHING, you need to wait for multiple Gen2s before you can know what to clean up.

I think it has merit (simple, more efficient, no tricky wrap-around issues).  

For what it is worth...

Reply:
>If you feel strongly that things could be completely out of control even within 10 seconds

It isn't that I'm concerned that we need to respond quickly, I'm more concerned about the need for constant callbacks / timers. My presumption is that it is better to take the cost of GetTicks() when you're using the pool than introducing a timer. I don't want to be spending time in the pool when you aren't actively using it or we're not already cleaning memory up.

I could roll back to not tracking the tls times and only trimming large tls under high memory pressure if we find GetTicks() to be too costly or just generally feel that tradeoff is worth it. (Or just track the times for those large buckets, etc.)

Reply:
Note that timers as not inefficient, they are actually efficient, because they are going off so rarely (just FYI, the threadpool has a timer that goes off every 30 seconds once the thread pool has started and no one cares). 

You are worried about the  fact that if the Pool is not used AT ALL you still have overhead.     That is easily addressed.   The most trivial way is to simply use Gen2s exclusively for activating Trim() like you do now.    The ageing works fine (it just means that you can't distinguish intervals of times that are shorter than a Gen2 GC.   That may slow you down A LITTLE, but frankly not enough to matter (as you point out, you already done something like this because yo only give back one array on every Trim).  

The question of whether to use real time or 'allocation activity' (eg. GCs) as the unit of measure, is an interesting one.    Each can have their place, and making a hybrid is not hard.   But I like my systems to be simple unless forced not to be, so I would be inclined to either 

1. Use GC2 soley as the Trim() activator OR
2. Use Timers soley as the Trim() activator.   (maybe with exponential backoff, but I would probably not add the complexity right off).

So why not (1)?

Reply:
One note on timers- I wouldn't want to be driven on a timer without putting the cleanup work on a worker thread. At that point I'd wonder why we weren't using the WaitForFullGCApproach method.

> you can't distinguish intervals of times that are shorter than a Gen2 GC

That could be quite some time, yes? If we get times only when Gen2 GCs happen we will always require 2 full Gen2 GCs to make an array eligible for collection (and a 3rd to actually collect). The cost of getting the environment tick count to cut that to 1(2) seems worthwhile, particularly if we don't track smaller buckets.  When we're going from idle to super busy / heavy allocating I'd like to see us respond right away when the first GC comes in.

Rolling over tick counts can cause a slight hiccup every month, but that didn't seem worth worrying about.

I'm totally ok with modifying our tick gathering for the thread locals. We could:

1. Don't gather times at all for the TLS, and simply purge the super large arrays if memory pressure is high. (Like over 16K length arrays)
    a. Super fast
    b. Frees up 17 ints per thread

or
 
2. Just gather times for the super large arrays and trim them as we are
    a. Faster
    b. Doesn't take any less memory

Reply:
> One note on timers- I wouldn't want to be driven on a timer without putting the cleanup work on a worker thread

I don't actually understand your concern here.  Timers call back on a thread-pool thread, which is perfectly fine (it gets returned to the pool when cleanup is done).   

Your idea of simply not tracking time for TLS is OK because it is simple (I would purge them uniformly as both the cost and benefit is proportional to size and thus there is not strong reason to prefer one over the other).   

However note that your concern that we don't react quickly is not clear to me.   Under high memory pressure, we should simply delete the cache entries independent of their age.   Thus under the case that matters most the algorithms are the same.   Indeed what makes this issue a 'must fix' is that it gets 'trapped' and never recovers.   Sure it takes one more GC for that to happen, but that was already true that it takes multiple GCs to 'clean up' and 10s of seconds to clean up the cache.   

Notice if you start with 'purge everything on high, otherwise do nothing' it is just a slight incremental step to 'just remember if this entry is old or new'  which allows you to do a better job when you are not under high memory pressure.     Then it is just another small step to doing that uniformly not only for the TLS case but for the per-processor case.   

Notice you can take the idea of not worrying about time one step further.   You could simply drop the TIcks/age field everywhere, and simply decide how aggressive to be based on memory pressure.  At high memory pressure ALWAYS remove everything  AT  Medium pressure remove some probability (say 50-50) of the TLS slots at random and half the per-processor slots, and low pressure, do nothing.   

This works, but I don't like it because at low memory pressure you will never clean up.   

This pushes me back to an age mechanism.   What I want is that if you don't get used, you get thrown away (eventually).  I don't mind if it takes two GCs to get there (we are already have to wait another GC because we free just after a GC.   

But I am OK with starting with a simpler mechanism (just rely on memory pressure).

Reply:
> I don't actually understand your concern here. Timers call back on a thread-pool thread, which is perfectly fine (it gets returned to the pool when cleanup is done).

Ignorance apparently :)

> I would purge them uniformly as both the cost and benefit is proportional to size and thus there is not strong reason to prefer one over the other

I'm not sure I follow this. Presuming we're still active, I'd assume we would want to prefer things that are both less likely to be active and have a larger impact on heap space.

> Under high memory pressure, we should simply delete the cache entries independent of their age.

I don't know that we want to do that. I suppose unnecessary allocations of small buffers (4K/16K) won't make a huge difference to memory pressure by adding unnecessary reallocations. Conversely, it also shouldn't make a huge pressure difference to hold these, but if we don't hold them we can definitely impact performance for a mainline scenario.

> This works, but I don't like it because at low memory pressure you will never clean up.

Precisely- which is why I'm tracking times. I think it is ok to only clear the stacks under low memory pressure, but I'd not want to remove that as well.

Right now we have one entry per thread per bucket in TLS. Thread termination will keep that from being unbounded, no? The largest bucket size with our current types (byte/char) is 1MB/2MB. Across threads and user types I can see the largest few buckets potentially easily eating up 100's of MB. With our our stacks, however, we're 8 deep per bucket per core (up to 64 cores). We're into GB territory there. If we're active, we're active, but if we're not there's no reason we shouldn't do work to mitigate getting into a high-pressure or OOM situation.

Reply:
> I'm not sure I follow this. Presuming we're still active, I'd assume we would want to prefer things that are both less likely to be active and have a larger impact on heap space.

My take is that the size of the buffer is uncorrelated with whether they are active, and BOTH their BENEFIT and RISK proportional to their size.  Thus I agree with you that you have to do something with the big buckets to have a useful effect, I am arguing that you should the small ones as well (why are they any different?)   Again this boils down to a simplicity argument.   Uniformity is simpler, so don't make distinctions (complicating the code), unless there is a good reason for it.  

But I think we are in agreement that we SHOULD take history into account (and then be uniform about it).   Note that the mechanism I am advocating is at worst slower (by 1 GC), but even that could be fixed (by adding a timer), so that it is BETTER (because you clean up promptly even when Gen 2 GCs are infrequent).   

But I think we are in agreement that we SHOULD take history into account (and then be uniform about it).   Note that the mechanism I am advocating is at worst slower (by 1 GC), but even that could be fixed (by adding a timer) (latency for the current implementation could be reduced in the same way by using timer)

I also agree that we do need to do something.   Both solutions will clean things up reasonably promptly so that is good.   In the end, I don't really mind the GetTickCount approach, I just find another approach better because it minimizes the impact to the most perf-critical part.   I can live with GetTIckCount however.

Reply:
> Both solutions will clean things up reasonably promptly so that is good. In the end, I don't really mind the GetTickCount approach, I just find another approach better because it minimizes the impact to the most perf-critical part. I can live with GetTIckCount however.

Long term I expect that we'll make a lot of changes to this. Given that we're so short on time I'm much more nervous about changing approaches and I'm pushing back a little harder than I normally would to help make sure we churn for strong reasons. :) Your feedback is _greatly_ appreciated. I've giving all of it a ton of thought.

Reply:
> TickCount is pretty fast at least on Windows (MeasureIt says it is < the cost of 3 call-returns). I don't know about Linux.

It is about 2x slower on Linux in a good case. It can be many times slower on older or less mainstream distros; or other Unix systems. https://github.com/dotnet/coreclr/issues/2257 has some context.